cpp/server/server.h

```diff
@@ -0,0 +1,340 @@
+#include "server/json_output.h"
+#include "server/proxy.h"
+#include "util/etcd.h"
+#include "util/periodic_closure.h"
+#include "util/read_key.h"
+#include "util/status.h"
+#include "util/thread_pool.h"
+#include "util/uuid.h"
+
+DEFINE_int32(node_state_refresh_seconds, 10,
+             "How often to refresh the ClusterNodeState entry for this node.");
+DEFINE_int32(watchdog_seconds, 120,
+             "How many seconds without successfully refreshing this node's "
+             "before firing the watchdog timer.");
+DEFINE_bool(watchdog_timeout_is_fatal, true,
+            "Exit if the watchdog timer fires.");
+
+namespace cert_trans {
+
+Gauge<>* latest_local_tree_size_gauge =
+    Gauge<>::New("latest_local_tree_size",
+                 "Size of latest locally generated STH.");
+
+
+template <class Logged>
+class Server {
+ public:
+  struct Options {
+    std::string server;
+    uint16_t port;
+
+    std::string etcd_root;
+
+    int num_http_server_threads;
+  };
+
+  static void StaticInit();
+
+  // Doesn't take ownership of anything.
+  Server(const Options& opts, std::shared_ptr<libevent::Base> event_base,
+         Database<Logged>* db, EtcdClient* etcd_client,
+         UrlFetcher* url_fetcher, LogSigner* log_signer,
+         CertChecker* cert_checker);
+  ~Server();
+
+  bool IsMaster() const;
+  MasterElection* election();
+  ConsistentStore<Logged>* consistent_store();
+  ClusterStateController<Logged>* cluster_state_controller();
+  LogLookup<Logged>* log_lookup();
+
+  void Initialise();
+  void Run();
+
+
+ private:
+  const Options& options_;
+  std::shared_ptr<libevent::Base> event_base_;
+  std::shared_ptr<libevent::EventPumpThread> event_pump_;
+  libevent::HttpServer http_server_;
+  Database<Logged>* const db_;
+  CertChecker* const cert_checker_;
+  const std::string node_id_;
+  UrlFetcher* url_fetcher_;
+  EtcdClient* etcd_client_;
+  MasterElection election_;
+  ThreadPool internal_pool_;
+  util::SyncTask server_task_;
+  StrictConsistentStore<Logged> consistent_store_;
+  std::unique_ptr<Frontend> frontend_;
+  std::unique_ptr<LogLookup<Logged>> log_lookup_;
+  std::unique_ptr<ClusterStateController<LoggedCertificate>>
+      cluster_controller_;
+  std::unique_ptr<ContinuousFetcher> fetcher_;
+  ThreadPool http_pool_;
+  JsonOutput json_output_;
+  std::unique_ptr<Proxy> proxy_;
+  std::unique_ptr<HttpHandler> handler_;
+  std::unique_ptr<std::thread> node_refresh_thread_;
+
+  DISALLOW_COPY_AND_ASSIGN(Server);
+};
+
+
+namespace {
+
+
+void RefreshNodeState(ClusterStateController<LoggedCertificate>* controller,
+                      util::Task* task) {
+  CHECK_NOTNULL(task);
+  const std::chrono::steady_clock::duration period(
+      (std::chrono::seconds(FLAGS_node_state_refresh_seconds)));
+  std::chrono::steady_clock::time_point target_run_time(
+      std::chrono::steady_clock::now());
+
+  while (true) {
+    if (task->CancelRequested()) {
+      task->Return(util::Status::CANCELLED);
```

Comment:
If you're adding cancellation, you should cancel the `alarm()` as well, otherwise you might have a surprise later!